Conversational LangChain model support

ishaan-mehta opened the pull request:
### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Adds support for conversational LangChain models (such as `RunnableWithMessageHistory`) by passing configurables to model invocation and calling `invoke` instead `__call__` for all models.

Adds check to ensure signature is not inferred from `lc_model` passed as a `str`.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Added support for prediction with any LangChain model loaded as a `python_function` (`pyfunc`) instead of select types. Added support  for passing `configurable` fields to `Runnable` LangChain models loaded as a `python_function` (`pyfunc`).

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

ishaan-mehta committed:
check for any Runnable in call_api()

Signed-off-by: Ishaan Mehta <45380942+ishaan-mehta@users.noreply.github.com>

ishaan-mehta committed:
add params as configurable

Signed-off-by: Ishaan Mehta <45380942+ishaan-mehta@users.noreply.github.com>

B-Step62 commented:
@ishaan-mehta Thank you so much for your contribution!!

Adding `configurable` parameter makes sense to me. However, accepting all Runnable types need to be a bit carefully done, because some types are not pickelable and need special saving logic. We have a list of current supported types [here](https://github.com/mlflow/mlflow/blob/2248511d5d86bbd751a0fea4241f78be7bc8f3c8/mlflow/langchain/utils.py#L143-L149), so would you mind testing other types you'd like to support and add them as test cases in [test_langchain_model_export.py](https://github.com/mlflow/mlflow/blob/2248511d5d86bbd751a0fea4241f78be7bc8f3c8/tests/langchain/test_langchain_model_export.py)?

ishaan-mehta committed:
only infer signature if actual model provided

Signed-off-by: Ishaan Mehta <45380942+ishaan-mehta@users.noreply.github.com>

ishaan-mehta commented:
Hi @B-Step62 — thanks for taking a look here! A couple thoughts:

Given that this check takes place in the prediction function where the model has already been saved and is now being loaded, does it matter which types are pickleable since we have already saved and loaded back the model? Any checks for whether the model is savable should take place in the `save_model()` function itself, no?

Also, with https://github.com/mlflow/mlflow/pull/11370, we are now able to log arbitrary LangChain model types as code (including ones we weren't previously able to save, such as `RunnableWithMessageHistory`), so I am not sure it makes sense to limit prediction to a select few types.

B-Step62 commented:
@ishaan-mehta Thank you so much for the explanation! I've overlooked that the condition check is in prediction function, and that makes sense to remove the limitation then🙂

B-Step62 commented on mlflow/langchain/__init__.py:
```diff
@@ -607,7 +607,9 @@ def predict(
         messages, return_first_element = self._prepare_messages(data)
-        results = process_api_requests(lc_model=self.lc_model, requests=messages)
+        results = process_api_requests(
+            lc_model=self.lc_model, requests=messages, configurables=params
```
Could you update the docstring of `params` argument to mention that it will passed as `config={"configurable": params}` to the underlying chain?

ishaan-mehta replied:
Good point, done

B-Step62 commented on mlflow/langchain/api_request_parallel_processor.py:
```diff
@@ -200,7 +201,10 @@ def call_api(
                             self.request_json,
-                            config={"callbacks": callback_handlers},
+                            config={
+                                "callbacks": callback_handlers,
+                                "configurable": self.configurables,
```
nit: can we define a `config` variable to avoid repetition?

ishaan-mehta replied:
Sure, done

B-Step62 reviewed (commented):
Overall change looks good to me, thank you so much for the contribution!

A few requests:
1. Could you run the pre-commit hook to format the code? you can run it by ``pre-commit run --all-files`` and it'll automatically update the files.
2. Could you add tests for (1) pass configurable as prediction parameter (2) run prediction with `RunnableWithMessageHistory`?

cc: @serena-ruan as SME

ishaan-mehta committed:
update docstring for params

Signed-off-by: Ishaan Mehta <45380942+ishaan-mehta@users.noreply.github.com>

ishaan-mehta committed:
update prediction logic to preserve previous behavior

Signed-off-by: Ishaan Mehta <45380942+ishaan-mehta@users.noreply.github.com>

ishaan-mehta committed:
add test for RunnableWithMessageHistory

Signed-off-by: Ishaan Mehta <45380942+ishaan-mehta@users.noreply.github.com>

ishaan-mehta committed:
add comment to test

Signed-off-by: Ishaan Mehta <45380942+ishaan-mehta@users.noreply.github.com>

ishaan-mehta commented:
Hi @B-Step62, I've addressed your comments, added a test that addresses prediction with a `RunnableWithMessageHistory` while passing configurables as `params`, and restructured my changes to `call_api()` so that the previous behavior (base types like `Retriever`, `Chain`, etc. being invoked with `return_only_outputs=True`) is preserved. Let me know if this looks good — thanks!

BenWilson2 commented:
@ishaan-mehta could you do a rebase to master? Let's see if we can get this merged in :)
